pandas/core/internals/concat.py

```diff
@@ -233,6 +227,24 @@ def dtype(self):
         else:
             return get_dtype(maybe_promote(self.block.dtype, self.block.fill_value)[0])
 
+    def is_valid_na_for(self, dtype: DtypeObj) -> bool:
+        """
+        Check that we are all-NA of a type/dtype that is compatible with this dtype.
+        """
+        if not self.is_na:
+            return False
+        if self.block is None:
+            return True
+
+        if self.dtype == object:
+            values = self.block.values
+            return all(
+                is_valid_nat_for_dtype(x, dtype) for x in values.ravel(order="K")
+            )
```

Comment:
This is not only required for object dtype, I think. Also float NaN is considered "all NaN" when it comes to ignoring the dtype in concatting dataframes:

```
In [39]: pd.concat([pd.DataFrame({'a': [np.nan, np.nan]}), pd.DataFrame({'a': [pd.Timestamp("2012-01-01")]})])
Out[39]: 
           a
0        NaT
1        NaT
0 2012-01-01

```

Reply:
the non-object case is handled below on L245-246.  or do you have something else in mind?

Reply:
Does my snippet above work with this PR? 
(if so, then I don't fully understand why the changes to `test_append_empty_frame_to_series_with_dateutil_tz` are needed)

Reply:
> Does my snippet above work with this PR?

yes it does

> (if so, then I don't fully understand why the changes to test_append_empty_frame_to_series_with_dateutil_tz are needed)

I think that's driven by something sketchy-looking in get_reindexed_values, will see if that can be addressed.

Reply:
better?